duolingo.py

```diff
@@ -28,20 +28,84 @@ class DuolingoException(Exception):
 class Duolingo(object):
     USER_AGENT = "Python Duolingo API/{}".format(__version__)
 
-    def __init__(self, username, password):
-        self.username = username
-        self.password = password
-        self.user_url = "https://duolingo.com/users/%s" % self.username
+    _url_base = "https://{host}{base_path}"
+    _url_user_no_auth = "/users?username={username}"
+    _url_login = "/login?fields="
+
+    def __init__(self, username="", password="", host="www.duolingo.com", base_path="/2017-06-30"):
+        self.host = host
+        self.base_path = base_path
+        self.base_url = self._url_base.format(host=self.host, base_path=self.base_path)
+
         self.session = requests.Session()
         self.leader_data = None
-        self.jwt = None
+        self.jwt = None        
+
+        if username:
+            self.username = username
+            if password:
+                self.password = password
+                self._login()
+            self._get_user_data(username)
+        self.voice_url_dict = None
 
-        if password:
-            self._login()
+    def _get_user_data(self, username):
+        """Gets user data with _get_data if logged in
+            and _get_user_data_no_auth otherwise
+        
+        Arguments:
+            username {str} -- username on Duolingo
+        """        
+        self.username = username
+        self.user_url = "https://duolingo.com/users/%s" % self.username # TODO: should be removed
+
+        data = self._get_data() if self.jwt else self._get_user_data_no_auth(username)
+        self.user_data = Struct(**data)
+
+    def _get_user_data_no_auth(self, username):
+        """Gets user data without authentication
+        
+        Arguments:
+            username {str} -- username on Duolingo
+        
+        Returns:
+            dict -- user data
+        """        
+        url = self.base_url + self._url_user_no_auth.format(username=username)
+        get = self._make_req(url).json()
+        if not "users" in get:
+            raise DuolingoException("users field not found in response")
+        if len(get["users"]) == 0:
+            raise DuolingoException("No users in response")
+        return get["users"][0]
+
+    def login(self, username, password):
```

Comment:
Can't this just call _login()?

Reply:
The idea is to separate initialisation and the login. The private method _login won't be necessary at the end.

Reply:
Yes, but we still need to support users using it the old way, we don't want to break backward compatibility.
I like the idea of adding `login()` as a new method but I think we need to support the old way, and therefore, if they're doing the same thing, you can just call one from the other? Or possibly remove `_login()` and have `login()` called from the init if username and password are given?

Reply:
I agree, I'll change that in the next commit.

I've actually noticed that there are two login methods:
`/login` and `/2017-06-30/login` 
and they behave differently and require different request data.

Duolingo website uses `/2017-06-30/login`. Should we use it?

Reply:
Actually I'm trying now `/2017-06-30/login` and it doesn't work from the module with self._make_req(), returns 403, although in Insomnia and Go the same URL with the same data works just fine. I can't figure out what is the problem.

Reply:
Oh, it's captcha. Unexpected. 
```json
{
    "blockScript": "/AsQJFZ9W/captcha/captcha.js?<...>",
    "vid": "",
    "jsRef": "",
    "hostUrl": "/AsQJFZ9W/xhr",
    "customLogo": null,
    "appId": "PXAsQJFZ9W",
    "uuid": "<...>",
    "logoVisibility": "hidden",
    "jsClientSrc": "/AsQJFZ9W/init.js",
    "firstPartyEnabled": "true",
    "refId": "<...>",
    "cssRef": ""
}
```

Reply:
Yeah, testing locally, and I just noticed the same thing. The `_login()` method is working and this new `login()` method does not. That is very strange.

Reply:
Ah, that captcha is frustrating, I wonder whz insomnia and Go doesn't hit that? I think I noticed in my earlier testing last month that having the user agent be the default for `requests` was causing issues sometimes, that's why I added the user agent static variable

Reply:
The `username` should be `identifier` like this:
```python
data = {"identifier": self.username, "password": self.password}
```

But it doesn't solve the captcha problem. We should probably raise a proper exception for that anyway.

Reply:
Setting user agent to `insomnia/7.0.6` solves it. What is the algorithm for triggering captcha...

Reply:
I still see a failure to login at that endpoint, even if I change the user agent and the data key from username to identifier.

Is it helpful to say there's a captcha in the exception, if there is no way to pass it to the user for solving?

Why can't we just use the endpoint the `_login()` method uses?

Reply:
Seems like it's not what is used in Duolingo apps, at least web app. But the easiest way now is to use `_login()` method.


Although, we should investigate captcha issue, in case the method we use is deprecated and be prepared.

Reply:
I guess they whitelisted certain user agents